File: backend/app/schemas/security.py
"""Pydantic schemas for security dashboard APIs."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field

from app.models.security import SecurityTicketStatus


class SecurityTicketResponse(BaseModel):
    """Security ticket payload."""

    id: UUID
    project_id: UUID | None = None
    title: str
    description: str
    status: SecurityTicketStatus
    location: str
    category: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class SecurityTicketUpdate(BaseModel):
    """Update payload for a security ticket."""

    status: SecurityTicketStatus = Field(..., description="Updated ticket status")


class ThreatData(BaseModel):
    """Threat score data."""

    entity_id: str | None = None
    headline_score: int = Field(..., ge=0, le=100)


class SecurityOverviewResponse(BaseModel):
    """Security dashboard overview payload."""

    facility_label: str | None = None
    project_id: UUID | None = None
    tickets: list[SecurityTicketResponse] = Field(default_factory=list)
    threat: ThreatData

File: frontend/src/api/security.ts
import type {
  Ticket,
  TicketStatus,
  ThreatData,
} from '../shared/components/app/security/types'
import { getJson, patchJson, type RequestOptions } from './utils/request'

interface SecurityOverviewResponse {
  facility_label?: string | null
  project_id?: string | null
  tickets?: Ticket[]
  threat?: ThreatData
}

interface SecurityTicketResponse {
  id: string
  project_id?: string | null
  title: string
  description: string
  status: TicketStatus
  location: string
  category: string
  created_at: string
  updated_at: string
}

export interface SecurityOverview {
  facilityLabel: string | null
  projectId: string | null
  tickets: Ticket[]
  threat: ThreatData
}

function mapTicket(payload: SecurityTicketResponse): Ticket {
  return {
    id: String(payload.id ?? '').trim(),
    title: payload.title ?? '',
    description: payload.description ?? '',
    status: payload.status,
    location: payload.location ?? '',
    category: payload.category ?? '',
  }
}

export async function getSecurityOverview(
  projectId?: string,
  options: RequestOptions = {},
): Promise<SecurityOverview> {
  const params = new URLSearchParams()
  if (projectId) {
    params.set('project_id', projectId)
  }
  const path = params.toString()
    ? `/api/v1/security/overview?${params.toString()}`
    : '/api/v1/security/overview'

  const payload = await getJson<SecurityOverviewResponse>(path, options)
  const threat = payload.threat ?? { entity_id: null, headline_score: 0 }
  return {
    facilityLabel: payload.facility_label ?? null,
    projectId: payload.project_id ?? null,
    tickets: (payload.tickets ?? []).map(mapTicket),
    threat,
  }
}

export async function updateSecurityTicketStatus(
  ticketId: string,
  status: TicketStatus,
): Promise<Ticket> {
  const payload = await patchJson<SecurityTicketResponse>(
    `/api/v1/security/tickets/${ticketId}`,
    { status },
  )
  return mapTicket(payload)
}
